### PR TITLE
Add canonical compiler examples

### PR DESCRIPTION
**Pull request by mhevery**
This PR is a beginning of canonical compiler examples (What do we expect the compiler to generate).

While doing this PR I ran into some issue and was forced to refactor:

1. Directives need to be passed into the `E` instructions since instantiation order can't be determined without breaking locality principle. Current implementation is a bit of a hack, follow on PR will create proper implementation.

2. The way queries register the local names is not consistent with the locality rule. In process of making new API, currently all of the query tests are broken. The local name `#local-name` on the element and the `exportAs` have to be resolved at runtime. /cc @pkozlowski-opensource

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -990,7 +1019,11 @@ export function lifecycle(lifeCycle: LifecycleHook, self?: any, method?: Functio
  * @param attrs The attrs attached to the container, if applicable
  */
 export function containerStart(
-    index: number, template?: ComponentTemplate<any>, tagName?: string, attrs?: string[],
+    index: number, 
+    directiveTypes?: DirectiveType<any>[], 
+    template?: ComponentTemplate<any>, 
+    tagName?: string, 
+    attrs?: string[],
     localName?: string): void {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: Shouldn't we unify this with `elementStart` (that is, if `localName` goes away from `elementStart` and is moved to attributes, it should go away from here as well, right?)

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -396,29 +396,35 @@ export function injectViewContainerRef(): ViewContainerRef {
...
  * @param attrs Statically bound set of attributes to be written into the DOM element on creation.
- * @param localName A name under which a given element is exported.
  *
  * Attributes are passed as an array of strings where elements with an even index hold an attribute
  * name and elements with an odd index hold an attribute value, ex.:
  * ['id', 'warning5', 'class', 'alert']
  */
 export function elementStart(
-    index: number, nameOrComponentDef?: string | ComponentDef<any>, attrs?: string[] | null,
-    localName?: string): RElement {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: Is your intention to merge `localName` into `attrs`? If so we will have to check for the `#` prefix at runtime, right?

Alternative approach would be to have `localNames: string[]` and have all the local names (element, component and directives) in there. This way we have clear distinction between attributes and local names.

File: packages/core/src/render3/assert.ts [hunk 3]
```diff
@@ -21,7 +21,10 @@
  * @returns The stringified value
  */
 function stringifyValueForError(value: any): string {
-  return typeof value === 'string' ? `"${value}"` : '' + value;
+  if (value && value.native && value.native.outerHTML) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/assert.ts]: Maybe create a dedicated function for stringifying `LNode`s instead?

File: packages/core/src/render3/instructions.ts [hunk 4]
```diff
@@ -466,9 +494,15 @@ function setUpAttributes(native: RElement, attrs: string[]): void {
   ngDevMode && assertEqual(attrs.length % 2, 0, 'attrs.length % 2');
   const isProceduralRenderer = (renderer as ProceduralRenderer3).setAttribute;
   for (let i = 0; i < attrs.length; i += 2) {
-    isProceduralRenderer ?
-        (renderer as ProceduralRenderer3).setAttribute !(native, attrs[i], attrs[i | 1]) :
-        native.setAttribute(attrs[i], attrs[i | 1]);
+    const key = attrs[i];
+    const value = attrs[i|1];
+    if (key.charAt(0) == '#') {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: Is your intention to have any code here or just skip attributes representing local names? If it is to skip, just flip the condition. Otherwise add info / comment / implementation here.

File: packages/core/src/render3/instructions.ts [hunk 5]
```diff
@@ -396,29 +396,35 @@ export function injectViewContainerRef(): ViewContainerRef {
  * Create DOM element. The instruction must later be followed by `elementEnd()` call.
  *
  * @param index Index of the element in the data array
- * @param nameOrComponentDef Name of the DOM Node or `ComponentDef`.
+ * @param nameOrComponentType Name of the DOM Node or `ComponentDef`.
  * @param attrs Statically bound set of attributes to be written into the DOM element on creation.
- * @param localName A name under which a given element is exported.
  *
  * Attributes are passed as an array of strings where elements with an even index hold an attribute
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: If your intention is to merge attributes and local names arrays, please update this part of the doc accordingly (but also please check my other comments on merging 2 collections)

File: packages/core/src/render3/instructions.ts [hunk 6]
```diff
@@ -831,56 +865,51 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
...
-    index: number, directive: T, directiveDef: DirectiveDef<T>, localName?: string): T;
-export function directive<T>(
-    index: number, directive?: T, directiveDef?: DirectiveDef<T>, localName?: string): T {
+export function directive<T>(index: number): T {
+  ngDevMode && assertDataInRange(index);
+  return data[index];
+}
+
+export function directiveCreate<T>(
+    index: number, directive: T, directiveDef: DirectiveDef<T>, localName?: string): T {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: I guess the `localName` is not the things here any more (logic to be moved to the `elementStart`?)

File: packages/core/src/render3/instructions.ts [hunk 7]
```diff
@@ -446,6 +455,25 @@ export function elementStart(
 
       if (attrs) setUpAttributes(native, attrs);
       appendChild(node.parent !, native, currentView);
+
+      if (hostComponentDef) {
+        // TODO(mhevery): This assumes that the directives come in correct order, which
+        // is not guaranteed. Must be refactored to take it into account.
+        directiveCreate(++index, hostComponentDef.n(), hostComponentDef);
+      }
+      if (directiveTypes) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: If we embed this logic in `E(...)` all of the sudden it needs to know about all the directives processing logic (instantiate, set inputs, do things in the order required by DI, ...) and local names logic. This will make a simple hello world app pulling more code than before, if I'm not mistaken. Did you check the impact of this change on the "Hello, World" app size?

An alternative would be to have another instruction (sth like `InstantiateDirectives([Directive1Type, Directive2Type, ...])`. I know that this would have minimal impact on real-life apps (as you need directives vast majority of the time) but still, would be nice to keep instructions small and independent (as much as possible...)

File: packages/core/src/render3/instructions.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: going with `string[]`

File: packages/core/src/render3/instructions.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: reverted

**Comment by mhevery** [comment]
The code currently contains `hack_declareDirectives` and `hack_findQueryName` functions which can't be correctly implemented until @kara fixes refactors the data structures and gives us a place to store these values.

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    it('should translate DOM structure', () => {
+      @Component({
+        selector: 'my-component',
+        template: `<div class="my-app" title="Hello">Hello <b>World</b>!</div>`
+      })
+      class MyComponent {
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['class', 'my-app', 'title', 'Hello'];
+      const MyComponentDef = defineComponent({
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/compiler_canonical_spec.ts]: Why the `MyComponentDef` isn't a static prop of `MyComponent` - is it intentional?

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  describe('elements', () => {
+    it('should translate DOM structure', () => {
+      @Component({
+        selector: 'my-component',
+        template: `<div class="my-app" title="Hello">Hello <b>World</b>!</div>`
+      })
+      class MyComponent {
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['class', 'my-app', 'title', 'Hello'];
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/compiler_canonical_spec.ts]: It will also need some file-wide index attached to the name as we can have multiple components in one file.

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+          factory: () => new SomeDirective(),
+        });
+      }
+
+      @Component({selector: 'my-component', template: `<child some-directive></child>`})
+      class MyComponent {
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['some-directive', ''];
+      const e0_dirs = [SomeDirective];
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/compiler_canonical_spec.ts]: Both `const` need some file-wide index in the name

File: packages/core/src/render3/instructions.ts [hunk 11]
```diff
@@ -427,30 +435,83 @@ export function elementStart(
...
+          ++index, directiveDef.n(), directiveDef, hack_findQueryName(directiveDef, localRefs));
+    }
+  }
+}
+
+/**
+ * This function returns the queryName for a directive. It is a hack since we should
+ * compute the query value only once and store it with the template (rather than on each invocation)
+ */
+function hack_findQueryName(
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: I know this is a "hack" and will go away, but could we rename it to something that doesn't have `query` in the name? Essentially what we are doing here is to connect local refs with directive instances so maybe `hack_mapLocalnameToDirectiveInstance` ?

File: packages/core/src/render3/instructions.ts [hunk 12]
```diff
@@ -427,30 +435,83 @@ export function elementStart(
...
+            -1, rendererFactory.createRenderer(native, hostComponentDef !.rendererType),
             ngStaticData));
       }
 
       // Only component views should be added to the view tree directly. Embedded views are
       // accessed through their containers because they may be removed / re-added later.
       node = createLNode(index, LNodeFlags.Element, native, componentView);
 
+      // TODO(misko): implement code which caches the local reference resolution
+      const queryName: string|null = hack_findQueryName(hostComponentDef, localRefs, '');
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: `queryName` => `localName` or `localNameForComponent`? Let's keep it independent from queries.

File: packages/core/src/render3/instructions.ts [hunk 13]
```diff
@@ -830,57 +905,47 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
  *        be created or retrieved out of order.
  * @param directive The directive instance.
  * @param directiveDef DirectiveDef object which contains information about the template.
+ * @param queryName Name under which the query can retrieve the directive instance.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: I would rename it to sht query independent, ex: `localName` or `nameInATemplate` - a name under which a directive instance is available in a template. Today query is making use of it name but it also declares names in a template and in the future we might have other usages (ex. animation queries).

tl;dr; my proposal is to keep this name query-independent.

File: packages/core/src/render3/instructions.ts [hunk 14]
```diff
@@ -830,57 +905,47 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
...
-    const diPublic = directiveDef !.diPublic;
-    if (diPublic) {
-      diPublic(directiveDef !);
+  if (index >= ngStaticData.length) {
+    ngStaticData[index] = directiveDef !;
+    if (queryName) {
+      ngDevMode &&
+          assertNotNull(previousOrParentNode.staticData, 'previousOrParentNode.staticData');
+      const nodeStaticData = previousOrParentNode !.staticData !;
+      (nodeStaticData.localNames || (nodeStaticData.localNames = [])).push(queryName, index);
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: I need to think a bit more about it, but I've got an impression that now we could skip saving  `localNames` is static data altogether. Previously we had to save a mapping between template-local names and directive instances since we haven't had this info on an element level (it was spread between element and directive instructions). But now everything is in one place.

So, maybe we can move logic of adding elements to queries in `elementStart` / `elementEnd`. This would let us store less info in memory.

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: changed

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: existing compiler has mechanism to deal with it. This is meant so we know what instructions to generate mainly.

File: packages/core/test/render3/compiler_canonical_spec.ts (hunk 10 above)
Comment (mhevery) [reply packages/core/test/render3/compiler_canonical_spec.ts]: 👍

File: packages/core/src/render3/instructions.ts (hunk 11 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: But that name is only used for `query` and nothing else. So I think the current name is right? Who else would use this value?

File: packages/core/src/render3/instructions.ts (hunk 12 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: see previous comment. I think this only makes sense for queries. Who else would use this information.

File: packages/core/src/render3/instructions.ts (hunk 13 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: see above

File: packages/core/src/render3/instructions.ts (hunk 14 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: follow up PRs welcome.

File: packages/core/src/render3/instructions.ts [hunk 15]
```diff
@@ -427,30 +435,83 @@ export function elementStart(
...
+ */
+function hack_declareDirectives(
+    index: number, directiveTypes: DirectiveType<any>[] | null | undefined,
+    localRefs: string[] | null | undefined, ) {
+  if (directiveTypes) {
+    // TODO(mhevery): This assumes that the directives come in correct order, which
+    // is not guaranteed. Must be refactored to take it into account.
+    for (let i = 0; i < directiveTypes.length; i++) {
+      // MEGAMORPHIC: `ngDirectiveDef` is a megamorphic property access here.
+      // This is OK, since we will refactor this code and store the result in `TViewState.data`
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: fixed

**Review by kara — APPROVED** [review]
LGTM
